pkg/smartcontract/zkpbinding/binding.go

```diff
@@ -0,0 +1,383 @@
+	if len(vkBytes) < verifyingKeyConstantPartLen {
+		return errors.New("unexpected len of constant-size part of serialized verifying key")
+	}
+	kvkLen := binary.BigEndian.Uint32(vkBytes[verifyingKeyConstantPartLen-4:])
+	if len(vkBytes) != verifyingKeyConstantPartLen+int(kvkLen*bls12381.SizeOfG1AffineCompressed) {
+		return errors.New("unexpected len of serialized verifying key")
+	}
+
+	var (
+		alphaG1Offset = 0
+		betaG2Offset  = bls12381.SizeOfG1AffineCompressed + bls12381.SizeOfG1AffineCompressed   // [α]1,[β]1
+		gammaG2Offset = bls12381.SizeOfG1AffineCompressed + bls12381.SizeOfG1AffineCompressed + // [α]1,[β]1
+			bls12381.SizeOfG2AffineCompressed // [β]2
+		deltaG2Offset = bls12381.SizeOfG1AffineCompressed + bls12381.SizeOfG1AffineCompressed + // [α]1,[β]1
+			bls12381.SizeOfG2AffineCompressed + bls12381.SizeOfG2AffineCompressed + // [β]2,[γ]2
+			bls12381.SizeOfG1AffineCompressed // [δ]1
+		kvkLenOffset = bls12381.SizeOfG1AffineCompressed + bls12381.SizeOfG1AffineCompressed + // [α]1,[β]1
+			bls12381.SizeOfG2AffineCompressed + bls12381.SizeOfG2AffineCompressed + // [β]2,[γ]2
+			bls12381.SizeOfG1AffineCompressed + bls12381.SizeOfG2AffineCompressed // [δ]1,[δ]2
+		kvkStartOffset = kvkLenOffset + 4
+	)
+	alphaG1 := vkBytes[alphaG1Offset : alphaG1Offset+bls12381.SizeOfG1AffineCompressed]
+	betaG2 := vkBytes[betaG2Offset : betaG2Offset+bls12381.SizeOfG2AffineCompressed]
+	gammaG2 := vkBytes[gammaG2Offset : gammaG2Offset+bls12381.SizeOfG2AffineCompressed]
+	deltaG2 := vkBytes[deltaG2Offset : deltaG2Offset+bls12381.SizeOfG2AffineCompressed]
+	kvks := make([][]byte, kvkLen)
+	for i := range kvks {
+		start := kvkStartOffset + i*bls12381.SizeOfG1AffineCompressed
+		end := start + bls12381.SizeOfG1AffineCompressed
+		kvks[i] = vkBytes[start:end]
+	}
+
+	// Generate verification contract from the template using the retrieved
+	// verification parameters.
+	tmpl := template.Must(template.New("generate").Funcs(template.FuncMap{
+		"byteSliceToStr": byteSliceToStr,
+	}).Parse(goVerificationTmpl))
+
+	err = tmpl.Execute(cfg.Output, tmplParams{
+		Alpha: alphaG1,
+		Beta:  betaG2,
+		Gamma: gammaG2,
+		Delta: deltaG2,
+		ICs:   kvks,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to generate template: %w", err)
+	}
+
+	if cfg.CfgOutput != nil {
+		_, err = cfg.CfgOutput.Write([]byte(verifyCfg))
+		if err != nil {
+			return fmt.Errorf("failed to generate contract configuration file: %w", err)
+		}
+	}
+	if cfg.GomodOutput != nil {
+		_, err = cfg.GomodOutput.Write([]byte(verifyGomod))
+		if err != nil {
+			return fmt.Errorf("failed to generate go.mod file: %w", err)
+		}
+	}
+	if cfg.GosumOutput != nil {
+		_, err = cfg.GosumOutput.Write([]byte(verifyGosum))
+		if err != nil {
+			return fmt.Errorf("failed to generate go.mod file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// byteSliceToStr is a codegen helper that converts byte slice to a go-like slice.
+func byteSliceToStr(s []byte) string {
```

Comment:
`%q` printf?

Reply:
Fixed.

Reply:
Not fixed, we need it to be presented as a go byte slice, i.e. like this:
```
[]byte{176, 235, 189, 253, 2, 48, 68, 109, 189, 147, 187, 42, 110, 114, 152, 166, 152, 89, 224, 228, 23, 101, 228, 96, 51, 152, 165, 110, 244, 50, 156, 173, 189, 104, 135, 11, 8, 41, 156, 186, 120, 1, 70, 1, 232, 79, 147, 196}
```

Reply:
What's the desired output here?